src/errors.rs

```diff
@@ -0,0 +1,508 @@
+#[non_exhaustive]
+pub enum FetchError {
+    #[error("Invalid URL for package: {url}")]
+    InvalidUrl {
+        url: String,
+        #[diagnostic(cause)]
+        error: url::ParseError,
+    },
+    #[error("Running as --frozen but needed to fetch {package}:{version}")]
+    Frozen {
+        package: PackageName,
+        version: Version,
+    },
+    #[error("Failed to unpack .crate at {}", src.display())]
+    Unpack {
+        src: PathBuf,
+        #[diagnostic(cause)]
+        error: UnpackError,
+    },
+    #[error("failed to open cached .crate at {}", target.display())]
+    OpenCached {
+        target: std::path::PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Download(#[from] DownloadError),
+}
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum UnpackError {
+    #[error("Failed to iterate archive")]
+    ArchiveIterate(#[diagnostic(cause)] std::io::Error),
+    #[error("Failed to read archive entry")]
+    ArchiveEntry(#[diagnostic(cause)] std::io::Error),
+    #[error("Invalid archive, {} wasn't under {}", entry_path.display(), prefix.to_string_lossy())]
+    InvalidPaths {
+        entry_path: PathBuf,
+        prefix: OsString,
+    },
+    #[error("Failed to unpack archive entry {}", entry_path.display())]
+    Unpack {
+        entry_path: PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error("Failed to finalize unpack to {}", target.display())]
+    LockCreate {
+        target: std::path::PathBuf,
+        #[diagnostic(cause)]
+        error: std::io::Error,
+    },
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+}
+
+//////////////////////////////////////////////////////////
+// FetchAuditError
+//////////////////////////////////////////////////////////
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum FetchAuditError {
+    // FIXME: would have to explicitly import URL for this error
+    #[error("invalid URL for foreign import {import_name} @ {import_url}")]
+    InvalidUrl {
+        import_name: ImportName,
+        import_url: String,
+        #[diagnostic(cause)]
+        error: url::ParseError,
+    },
+    #[diagnostic(transparent)]
+    #[error(transparent)]
+    Download(#[from] DownloadError),
+    #[diagnostic(transparent)]
+    #[error(transparent)]
+    Toml(#[from] TomlParseError),
+    #[diagnostic(transparent)]
+    #[error(transparent)]
+    Validate(#[from] StoreValidateErrors),
+}
+
+//////////////////////////////////////////////////////////
+// DownloadError
+//////////////////////////////////////////////////////////
+
+#[derive(Debug, Error, Diagnostic)]
+#[non_exhaustive]
+pub enum DownloadError {
+    #[error("failed to acquire permission to download {url}")]
+    DonwloadDenied {
+        url: reqwest::Url,
+        #[diagnostic(cause)]
+        error: tokio::sync::AcquireError,
+    },
```

Comment:
This should be impossible as we never close the semaphore. Perhaps we should just unwrap that error?

I worry that this message is misleading as-is right now if it did fire, because "permission to download" sounds like perhaps a HTTP 403 Forbidden or filesystem permission error, which it isn't.